### ISSUE TITLE
Investigating the mouse-hover problem in emscripten nuklear+Oryol 

### ISSUE THREAD

**Issue by floooh** [vurtun/nuklear#207]
I'm currently debugging the mouse-hover problem I'm seeing in my integration of nuklear for Oryol here: http://floooh.github.io/oryol-samples/asmjs/NuklearUIAdvanced.html, and I'm opening this bug to record my findings, may be at one point you can jump in and point me in the right direction :)

The buttons in the left most window, starting at the second row, don't get 'hovered' when the mouse is over them. Moving the window to the middle fixes the problem, but when moving towards the right, the problem happens again. This happens only in emscripten, so I'm limited to printf-debugging.

Here's what I found out so far:

The button doesn't get hovered because: in nk_button_behaviour(), the incoming nk_input*  is null, moving up the callstack to nk_button_text(), this will happen when the widget is in state NK_WIDGET_ROM, which seems to be the 'partially visible' (or occluded?) state...

Another interesting thing: up in the 'button_demo()' function, I have commented everything out except that first Push Button, and it looks like this:

```cpp
    ui_widget(ctx, media, 35);
    if (nk_button_label(ctx, "Push me"))
        fprintf(stdout, "pushed!\n");
```

The ui_widget() function looks like this:
```cpp
static void
ui_widget(struct nk_context *ctx, struct media *media, float height)
{
    static const float ratio[] = {0.15f, 0.85f};
    nk_style_set_font(ctx, &media->font_22->handle);
    nk_layout_row(ctx, NK_DYNAMIC, height, 2, ratio);
    nk_spacing(ctx, 1);
}
```
If I'm commenting out the call to nk_spacing(ctx, 1), the button becomes smaller, and the mouse hover works in emscripten.

That is all so far, I'll try to find out now why the button is in NK_WIDGET_ROM state.

Btw, how does nuklear detect the current canvas size, in Imgui I'm explicitely setting this at the start of a frame, but I haven't seen something similar in nuklear. I'm mentioning this because this might be a difference between emscripten and other platforms in Oryol, the size of the WebGL canvas might change, if nuklear uses the 'display resolution/canvas size', this might one potential cause for the problem.

**Comment by vurtun** [comment]
OK this is strange question but do you really have the newest version. Because I fixed almost exactly that bug yesterday but for `nk_layout_row_begin`, `nk_layout_row_push` and `nk_layout_row_end` . The problem was that I calculated the size of the widgets wrong and the button went out of the window and therefore could not be pressed. I will do some tests and see if I can find something out. Problem is that it works for me so I don't know how succesful I will be.

As for the canvas size. I do not track it at the moment because I did not need it so far for anything.

**Comment by floooh** [comment]
Now it gets strange, I've added printfs to nk_widget() which returns the widget state. When running natively in OSX, everything is alright:
```
clip:   60.00, 96.00, 225.00, 540.00
bounds: 97.15, 100.00, 187.85, 35.00
  NK_WIDGET_VALID
```
The exact same values on emscripten, yield NK_WIDGET_ROM, meaning the if (NK_CONTAINS...) triggers:
<img width="254" alt="screen shot 2016-08-08 at 19 50 48" src="https://cloud.githubusercontent.com/assets/1699414/17490152/bd25537e-5da1-11e6-9643-63691c52db1f.png">

I've played around with the fairly complex looking macros under NK_CONTAINS(), wrapped them into functions, etc... nothing so far. Digging deeper...

**Comment by floooh** [comment]
This is the commit I'm working with:
```
commit a0870bd30723c687630b119ec75e30b4ec0e9c1a
Author: vurtun <polygone@gmx.net>
Date:   Sun Aug 7 21:53:20 2016 +0200

    Fixed bug in layout API calculating wrong width

    The dynamic immediate mode layout API with `nk_layout_row_begin`,
    `nk_layout_row_push` and `nk_layout_row_end` caused an infinite
    scrollbar bug by wrongfully calculating the required widget sizes.
    This bug is now fixed and everything is correctly calculated.
```

**Comment by floooh** [comment]
More interesting findings: the state flips between ROM and VALID at button X coordinate 128 and 512:

ROM at x=127:
<img width="249" alt="screen shot 2016-08-08 at 19 58 20" src="https://cloud.githubusercontent.com/assets/1699414/17490396/ea20d960-5da2-11e6-99ba-9bd7b78c604b.png">

Valid at x=128:
<img width="258" alt="screen shot 2016-08-08 at 19 58 42" src="https://cloud.githubusercontent.com/assets/1699414/17490414/fdd19260-5da2-11e6-9e10-73e0ea019353.png">

Valid at x=511:
<img width="252" alt="screen shot 2016-08-08 at 19 59 16" src="https://cloud.githubusercontent.com/assets/1699414/17490429/0ab6748c-5da3-11e6-9402-7c357be88945.png">

And then ROM again at 512:
<img width="268" alt="screen shot 2016-08-08 at 20 00 04" src="https://cloud.githubusercontent.com/assets/1699414/17490436/18dd72b8-5da3-11e6-99cf-c31aff5bdf6a.png">

Since the values are float this is a 'bit' strange :) Worst case is that this is a code-gen bug in emscripten, but I haven't encountered such strange behaviour so far. I'll meditate a bit more over this...

**Comment by vurtun** [comment]
Yeah that is the correct `commit`. I don't know much about emscript, but is there a possibility that because of floating point error (in contrast to window ratio in the array provided by `nk_layout_row`) could expand the widget width and bloat it up by one or more pixel and therefore cause the widget to be only partially visible and not updatable? That may explain why it works for native but not emscript.

**Comment by floooh** [comment]
I think all floating point computation is done in double precision in emscripten, since this is the Javascript floating point type... not sure how this could cause the problem though.

**Comment by floooh** [comment]
...I'll continue with a minimal test case, just the macros and values from above. Makes it easier to meditate over the generated code :)

**Comment by floooh** [comment]
Ok, semi-success :) I can reproduce the problem in native builds when doing the comparison with doubles, e.g. in nk_widget():

```cpp
double cx = c->x;
double cy = c->y;
double cw = c->w;
double ch = c->h;
double bx = bounds->x;
double by = bounds->y;
double bw = bounds->w;
double bh = bounds->h;

    if (!NK_INTERSECT(cx, cy, cw, ch, bx, by, bw, bh)) {
        printf("  NK_WIDGET_INVALID\n");
        return NK_WIDGET_INVALID;
    }
    if (!NK_CONTAINS(bx, by, bw, bh, cx, cy, cw, ch)) {
        printf("  NK_WIDGET_ROM\n");
        return NK_WIDGET_ROM;
    }
    printf("  NK_WIDGET_VALID\n");
    return NK_WIDGET_VALID;
```
This causes the exact same behaviour on my native OSX build as the emscripten build :)

**Comment by floooh** [comment]
I think the problem is the less-equal check in the macro NK_BETWEEN, when I change the comparisons in there  to 'less' instead of 'less-equal', all kinds of strange things happen (text isn't rendered, hover isn't detected...), so the check seems to depend on one floating pointer number being exactly equal to another, which might explain why the behaviour is different with higher floating point precision.

**Comment by vurtun** [comment]
OK I am able to reproduce it as well. But I still persist that it is an floating point error. While converting from float to double from `bounds.x` to `bx` it gains `0.00000002587906` and `bx + bw` is `230.00000762...` while `bounds.x + bounds.w` is a perfect `230`.

**Comment by floooh** [comment]
One fairly heavy-handed fix which works both native and in emscripten is to simply force the comparisons to integer, e.g. this works (in nk_widget()), I've running the whole demo-app again and don't notice any strange behaviour or see any rendering errors (both native OSX and emscripten):

```cpp
int cx = c->x;
int cy = c->y;
int cw = c->w;
int ch = c->h;
int bx = bounds->x;
int by = bounds->y;
int bw = bounds->w;
int bh = bounds->h;

    if (!NK_INTERSECT(cx, cy, cw, ch, bx, by, bw, bh))
        return NK_WIDGET_INVALID;
    if (!NK_CONTAINS(bx, by, bw, bh, cx, cy, cw, ch))
        return NK_WIDGET_ROM;
    return NK_WIDGET_VALID;
```

**Comment by vurtun** [comment]
Yeah probably the right approach. I don't think there are any downsides here since it really doesn't matter to have additional precision here.

**Comment by vurtun** [comment]
OK I fixed the issue by pushing the new commit a384c6ed80d93603389495e03b2902ef5c259f17. Hopefully this is now fixed.

**Comment by floooh** [comment]
Looking good, thanks a lot :) I'll upload new samples to the demo URL soon-ish. There's 2 things that I still need to fix in my integration: keyboard input, and images currently don't get mipmaps generated.
